File: static/torn/member-report.js
/* Copyright (C) 2021-2023 tiksan

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU Affero General Public License as published
by the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU Affero General Public License for more details.

You should have received a copy of the GNU Affero General Public License
along with this program.  If not, see <https://www.gnu.org/licenses/>. */

const tzOffset = new Date().getTimezoneOffset();

$(document).ready(function () {
    function addReport(item) {
        let container = $("#existing-report-container");

        if (container.text().includes("No results found...")) {
            $("#existing-report-container").empty();
        }

        container.append(
            $("<div>", {
                class: "card mx-2 mt-2 report-card",
                "data-report-id": item.report_id,
            }).append(
                $("<div>", {
                    class: "row p-2",
                }).append([
                    $("<div>", {
                        class: "col-sm-12 col-md-4 col-xl-4 mt-1",
                        text: `${item.faction.name} [${item.faction.id}]`,
                    }),
                    $("<div>", {
                        class: "col-sm-12 col-md-6 col-xl-4 mt-1",
                        text: `${new Date(
                            item.start_timestamp * 1000 - tzOffset
                        ).toDateString()} to ${new Date(
                            item.end_timestamp * 1000 - tzOffset
                        ).toDateString()}`,
                    }),
                    $("<div>", {
                        class: "col-sm-12 col-md-2 col-xl-4 mt-1",
                    }).append(
                        $("<div>", {
                            class: "w-100 justify-content-end d-flex",
                        }).append([
                            $("<button>", {
                                class: "btn btn-sm btn-outline-secondary view-report me-2",
                                "data-report-id": item.report_id,
                            }).append(
                                $("<i>", {
                                    class: "fa-regular fa-eye",
                                })
                            ),
                            $("<button>", {
                                class: "btn btn-sm btn-outline-secondary delete-report",
                                "data-report-id": item.report_id,
                            }).append(
                                $("<i>", {
                                    class: "fa-regular fa-trash-can",
                                })
                            ),
                        ])
                    ),
                ])
            )
        );
    }

    function deleteReport() {
        const rid = $(this).attr("data-report-id");
        const xhttp = new XMLHttpRequest();

        xhttp.onload = function () {
            let response = xhttp.response;

            if (response.code === 1) {
                $(`.report-card[data-report-id="${rid}"]`).remove();
            }
        };

        xhttp.responseType = "json";
        xhttp.open("DELETE", `/api/report/faction/members/${rid}`);
        xhttp.setRequestHeader("Content-Type", "application/json");
        xhttp.send();
    }

    function viewReport() {
        const rid = $(this).attr("data-report-id");
        window.location.href = `/torn/factions/member-report/${rid}`;
    }

    const xhttp = new XMLHttpRequest();

    xhttp.onload = function () {
        let response = xhttp.response;

        if ("code" in response) {
            generateToast(
                "Reports Load Failed",
                `The Tornium API server has responded with \"${response["message"]}\" to the submitted request.`
            );
            $("#existing-report-container")
                .empty()
                .append(
                    $("<i>", {
                        class: "fa-solid fa-skull-crossbones",
                        text: "Data failed to load...",
                    })
                );
            return;
        }

        if (response.count > 0) {
            $("#existing-report-container").empty();
            $("#report-count").text(commas(response.count));
            $.each(response.reports, function (index, item) {
                addReport(item);
            });

            $(".delete-report").on("click", deleteReport);
            $(".view-report").on("click", viewReport);
        }
    };

    xhttp.responseType = "json";
    xhttp.open("GET", `/api/report/faction/members`);
    xhttp.setRequestHeader("Content-Type", "application/json");
    xhttp.send();

    $("#generate-report").on("click", function () {
        $("#generate-report").attr("disabled", true);

        const faction = $("#faction-id-input").val();
        let startTime = Math.ceil(
            (new Date($("#start-time-input").val()) - tzOffset) / 1000
        );
        let endTime = Math.ceil(
            (new Date($("#end-time-input").val()) - tzOffset) / 1000
        );
        let availability = $('input[name="data-availability"]:checked').val();

        if (availability === undefined) {
            availability = "user";
        }

        const selectedOptions = $("#ps-select").find(":selected");
        let selectedPS = [];

        $.each(selectedOptions, function (index, item) {
            selectedPS.push(item.text);
        });

        if (
            faction === "" ||
            selectedPS.length === 0 ||
            (availability !== "user" && availability !== "faction")
        ) {
            generateToast("Invalid Input", "An inputted value was not valid");
            $("#generate-report").attr("disabled", false);
            return;
        } else if (endTime - startTime < 3600 * 24 * 2) {
            generateToast(
                "Invalid Dates",
                "The selected dates must be two or more days apart"
            );
            $("#generate-report").attr("disabled", false);
            return;
        } else if (startTime <= 0 || endTime <= 0) {
            generateToast(
                "Invalid Time",
                "An inputted date time was not valid"
            );
            $("#generate-report").attr("disabled", false);
            return;
        }

        const xhttp = new XMLHttpRequest();

        xhttp.onload = function () {
            let response = xhttp.response;
            $("#generate-report").attr("disabled", false);

            if ("code" in response) {
                generateToast(
                    "Report Generation Failed",
                    `The Tornium API server has responded with \"${response["message"]}\" to the submitted request.`
                );
                return;
            }

            addReport(response);
            $(".delete-report").on("click", deleteReport);
            $(".view-report").on("click", viewReport);
            $("#faction-id-input").val("");
            $("#start-time-input").val("");
            $("#end-time-input").val("");
        };

        xhttp.responseType = "json";
        xhttp.open("POST", "/api/report/faction/members");
        xhttp.setRequestHeader("Content-Type", "application/json");
        xhttp.send(
            JSON.stringify({
                faction_id: faction,
                start_time: startTime,
                end_time: endTime,
                availability: availability,
                selected_stats: selectedPS,
            })
        );
    });
});

File: celery/tornium_celery/tasks/reports/member_report.py
# Copyright (C) 2021-2023 tiksan
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU Affero General Public License as published
# by the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Affero General Public License for more details.
#
# You should have received a copy of the GNU Affero General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.

import datetime
import logging
import time
import typing

import celery
from celery.utils.log import get_task_logger
from peewee import DoesNotExist
from tornium_commons import rds
from tornium_commons.models import Faction, MemberReport, PersonalStats, User

from ..api import tornget

logger: logging.Logger = get_task_logger("celery_app")


@celery.shared_task(
    name="tasks.reports.member_ps_error",
    routing_key="quick.reports.member_ps_error",
    queue="quick",
)
def member_ps_error(request, exc, traceback, rid: str):
    try:
        report: MemberReport = MemberReport.get_by_id(rid)
    except DoesNotExist:
        return

    report.status = 999
    report.save()


@celery.shared_task(
    name="tasks.reports.member_ps_success",
    routing_key="quick.reports.member_ps_success",
    queue="quick",
)
def member_ps_success(rid: str):
    try:
        report: MemberReport = MemberReport.get_by_id(rid)
    except DoesNotExist:
        return

    if report.status == 1:
        return

    report.status = 2
    report.save()


@celery.shared_task(
    name="tasks.reports.enqueue_member_ps",
    routing_key="default.reports.enqueue_member_ps",
    queue="default",
)
def enqueue_member_ps(faction_data: dict, api_keys: tuple, rid: str):
    if len(api_keys) == 0:
        raise ValueError("No API calls provided")

    try:
        Faction.insert(
            tid=faction_data["ID"],
            name=faction_data["name"],
            tag=faction_data["tag"],
            respect=faction_data["respect"],
            capacity=faction_data["capacity"],
            leader=faction_data["leader"],
            coleader=faction_data["co-leader"],
        ).on_conflict(
            conflict_target=[Faction.tid],
            preserve=[
                Faction.name,
                Faction.tag,
                Faction.respect,
                Faction.capacity,
                Faction.leader,
                Faction.coleader,
            ],
        ).execute()
    except Exception as e:
        logger.exception(e)

    call_count = 0

    try:
        report: MemberReport = MemberReport.get_by_id(rid)
    except DoesNotExist:
        logger.debug(f"No report found - {rid}")
        raise ValueError("Illegal RID")

    requested_stats = ",".join(tuple(report.requested_data))

    for member_id, member_data in faction_data["members"].items():
        # Max 25 calls per key per minute
        try:
            User.insert(
                tid=int(member_id),
                name=member_data["name"],
                level=member_data["level"],
                last_refresh=datetime.datetime.utcnow(),
                faction=faction_data["ID"],
                status=member_data["last_action"]["status"],
                last_action=datetime.datetime.fromtimestamp(member_data["last_action"]["timestamp"]),
            )
        except Exception as e:
            logger.exception(e)
            raise e

        # If the member has joined since the report's start timestamp
        # Not always effective if the member has been in the faction before
        if member_data["days_in_faction"] < (int(time.time()) - report.start_timestamp) / (3600 * 24):
            from_ts = int(time.time()) - member_data["days_in_faction"] * 24 * 3600
        else:
            from_ts = report.start_timestamp

        logger.info(
            f"User {member_data['name']} [{member_id}] - from {from_ts} - countdown {60 * (call_count // (len(api_keys) * 25))}"
        )
        tornget.signature(
            kwargs={
                "endpoint": f"user/{member_id}?selections=personalstats&stat={requested_stats}&timestamp={from_ts}",
                "key": api_keys[call_count % len(api_keys)],
            }
        ).apply_async(
            countdown=60 * (call_count // (len(api_keys) * 25)),
            link=store_member_ps.signature(
                kwargs={
                    "tid": int(member_id),
                    "rid": rid,
                    "timestamp": from_ts,
                },
                queue="quick",
            ),
        )
        call_count += 1

    for member_id, member_data in faction_data["members"].items():
        logger.info(
            f"User {member_data['name']} [{member_id}] - to {report.end_timestamp} - countdown {60 * (call_count // (len(api_keys) * 25) + 1)}"
        )
        tornget.signature(
            kwargs={
                "endpoint": f"user/{member_id}?selections=personalstats&stat={requested_stats}&timestamp={report.end_timestamp}",
                "key": api_keys[call_count % len(api_keys)],
            }
        ).apply_async(
            countdown=60 * (call_count // (len(api_keys) * 25) + 1),
            link=store_member_ps.signature(
                kwargs={
                    "tid": int(member_id),
                    "rid": rid,
                    "timestamp": report.end_timestamp,
                },
                queue="quick",
            ),
        )
        call_count += 1

    member_ps_success.apply_async(
        countdown=60 * (call_count // (len(api_keys) * 25) + 2),
        kwargs={
            "rid": rid,
        },
    )

    rds().set(
        f"tornium:report:{rid}:endtime",
        int(time.time()) + 60 * (call_count // (len(api_keys) * 25) + 1),
        nx=True,
        ex=60 * (call_count // (len(api_keys) * 25) + 3),
    )


@celery.shared_task(
    name="tasks.reports.store_member_ps",
    routing_key="quick.reports.store_member_ps",
    queue="quick",
)
def store_member_ps(member_data: dict, tid: int, rid: str, timestamp: int):
    # timestamp must be in UTC

    try:
        if member_data.get("personalstats") is None:
            raise ValueError("Invalid API data")

        try:
            report: MemberReport = MemberReport.get_by_id(rid)
        except DoesNotExist:
            raise ValueError("Illegal RID")

        if report.start_timestamp > timestamp > report.end_timestamp:
            raise ValueError("Illegal timestamp")

        timestamp = int(
            datetime.datetime.fromtimestamp(timestamp)
            .replace(minute=0, second=0, tzinfo=datetime.timezone.utc)
            .timestamp()
        )

        if report.end_timestamp == timestamp:
            logger.info(f"Storing {tid} for {timestamp} as end timestamp")
        else:
            logger.info(f"Storing {tid} for {timestamp} as start timestamp")

        pid = int(bin(tid << 8), 2) + int(bin(timestamp), 2)

        try:
            PersonalStats(
                pstat_id=pid,
                tid=tid,
                timestamp=int(time.time()),
                ps_data=member_data["personalstats"],
            ).save()
        except Exception as e:
            if timestamp == report.end_timestamp:
                report.end_ps[str(tid)] = None
            else:
                report.start_ps[str(tid)] = None

            raise e

        if timestamp == report.end_timestamp:
            report.end_ps[str(tid)] = pid
        else:
            report.start_ps[str(tid)] = pid

        if report.status == 0:
            report.status = 1

        now = datetime.datetime.utcnow()

        if report.last_updated < now:
            report.last_updated = now

        report.save()
    except Exception as e:
        logger.exception(e)
        raise e
